test/shared/src/main/scala-dotty/zio/test/AssertionVariants.scala

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.test.Assertion.Render._
+
+trait AssertionVariants {
+
+  /**
+   * Makes a new assertion that requires a value equal the specified value.
+   */
+  final def equalTo[A](expected: A): Assertion[A] =
```

Comment:
Can we use the new `Eql` in Dotty?

Reply:
Unfortunately not, or at least not right now. We have an issue with the type of assertions composed with combinators such as `||` being inferred as `Nothing` when we use the same technique. See minimized example and link below. This looked like a type checking issue but I wanted to look at it a little more before opening an issue with the Dotty team.

```scala
object Example extends App {
  
  def assert[A](value: A)(assertion: Assertion[A]): Boolean =
    assertion.run(value)
  
  trait Assertion[-A] { self =>
    def run(a: A): Boolean
    def ||[A1 <: A](that: Assertion[A1]): Assertion[A1] =
      a => self.run(a) && that.run(a)
  }
  
  def equalTo[A, B](expected: A)(implicit eq: Eql[A, B]): Assertion[B] =
    _ == expected
  
  def isGreaterThan(a: Int): Assertion[Int] =
    _ > a
  
  // okay
  assert(List(1, 2, 3))(equalTo(List.empty))
  
  // does not compile
  // values of type Int and String cannot be compared with ==  or != 
  // assert("Adam")(equalTo(1))
  
  // does not compile
  // expected Assertion[Int] but found Assertion[Nothing]
  // assert(10)(equalTo(5) || isGreaterThan(9))
}
```

https://scastie.scala-lang.org/p41x4qlvQmiXbxyJKk7K9w

Reply:
I thought it was the `A1 <: A` that's forcing it to be `Nothing`, but changing that has no effect. Interesting...

We can make "too much" compile but still use `Eql` by adding a type bound:

```scala
  def equalTo[A, B >: A](expected: A)(implicit eq: Eql[A, B]): Assertion[B] =
    _ == expected
```

I'm not sure if using `Eql` would have any practical benefits in the short-term so maybe it's not worth it. Thoughts?

Reply:
I noticed the same thing too and initially rejected it because it would allow `assert("Adam")(equalTo(1))` to compile so it didn't seem like it was buying us anything. But thinking about it more I think it would allow us to at least unify the signature of `equalTo` and move it back into `Assertion.scala`, removing the `AssertionVariants` files, which would be nice.

Reply:
Ah, with the type bound on Scala 2 comparison's that don't make sense compile, so we can't do that. I don't think we really get any benefit out of using `Eql` in Dotty until we resolve the compilation issue but am fine either way.

Reply:
OK, no worries